includes/class-wcs-batch-processing-controller.php

```diff
@@ -0,0 +1,572 @@
+<?php
+/**
+ * This class is a helper intended to handle data processings that need to happen in batches in a deferred way.
+ * It abstracts away the nuances of (re)scheduling actions and dealing with errors.
+ *
+ * Usage:
+ *
+ * 1. Create a class that implements WCS_Batch_Processor.
+ *
+ * 2. Whenever there's data to be processed invoke the 'enqueue_processor' method in this class,
+ *    passing the class name of the processor.
+ *
+ * That's it, processing will be performed in batches inside scheduled actions; enqueued processors will only
+ * be dequeued once they notify that no more items are left to process (or when `force_clear_all_processes` is invoked).
+ * Failed batches will be retried after a while.
+ *
+ * This is heavily inspired by core's version at Automattic\WooCommerce\Internal\BatchProcessing\BatchProcessingController.
+ *
+ * @package  WooCommerce Subscriptions
+ * @category Class
+ * @since    x.x.x
+ */
+class WCS_Batch_Processing_Controller {
+	/*
+	 * Identifier of a "watchdog" action that will schedule a processing action
+	 * for any processor that is enqueued but not yet scheduled
+	 * (because it's been just enqueued or because it threw an error while processing a batch),
+	 * that's one single action that reschedules itself continuously.
+	 */
+	const WATCHDOG_ACTION_NAME = 'wcs_schedule_pending_batch_processes';
+
+	/*
+	 * Identifier of the action that will do the actual batch processing.
+	 * There's one action per enqueued processor that will keep rescheduling itself
+	 * as long as there are still pending items to process
+	 * (except if there's an error that caused no items to be processed at all).
+	 */
+	const PROCESS_SINGLE_BATCH_ACTION_NAME = 'wcs_run_batch_process';
+
+	const ENQUEUED_PROCESSORS_OPTION_NAME = 'wcs_pending_batch_processes';
+	const ACTION_GROUP                    = 'wcs_batch_processes';
+	const LOGS_CONTEXT                    = 'wcs-batch-processing';
+
+	/**
+	 * Maximum number of failures per processor before it gets dequeued.
+	 */
+	const FAILING_PROCESS_MAX_ATTEMPTS_DEFAULT = 5;
+
+	/**
+	 * Instance of WC_Logger class.
+	 *
+	 * @var \WC_Logger_Interface
+	 */
+	private $logger;
+
+	/**
+	 * Singleton instance.
+	 *
+	 * @var WCS_Batch_Processing_Controller
+	 */
+	private static $instance;
+
+	/**
+	 * Constructor.
+	 *
+	 * Schedules the necessary actions to process batches.
+	 */
+	private function __construct() {
+		add_action(
+			self::WATCHDOG_ACTION_NAME,
+			function () {
+				$this->handle_watchdog_action();
+			}
+		);
+
+		add_action(
+			self::PROCESS_SINGLE_BATCH_ACTION_NAME,
+			function ( $batch_process ) {
+				$this->process_next_batch_for_single_processor( $batch_process );
+			},
+			10,
+			2
```

Comment:
This function is only accepting 1 arg isn't it? `$batch_process`

Reply:
I think it was copied almost verbatim [from WC core](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php#L74-L81) by you, @xristos3490  :) 
And yes, you're right @james-allan , there is just one parameter.

Reply:
True that :) So, the reasoning for the closure is to avoid removals :)

Reply:
Yeah that's ok. We generally don't like using closures like this but given this is a copy of the WC Core file, we can live with that. :)